tests/unit/test-class-wc-payment-gateway-wcpay.php

```diff
@@ -2338,6 +2355,287 @@ public function test_no_payment_is_processed_for_woopay_preflight_check_request(
+	}
+
+	public function test_should_use_new_process_determines_positive_woopay_payment() {
+		// The new payment process is only accessible in dev mode.
+		WC_Payments::mode()->dev();
+
+		$order = WC_Helper_Order::create_order();
+
+		$_POST['platform-checkout-intent'] = 'pi_ZYX';
+
+		$this->expect_new_payment_process_factor( Factor::WOOPAY_PAYMENT(), true );
+		$this->wcpay_gateway->should_use_new_process( $order );
+	}
+
+	public function test_should_use_new_process_determines_negative_woopay_payment() {
+		// The new payment process is only accessible in dev mode.
+		WC_Payments::mode()->dev();
+
+		$order = WC_Helper_Order::create_order();
+
+		// phpcs:ignore WordPress.Security.NonceVerification.Missing
+		unset( $_POST['platform-checkout-intent'] );
+
+		$this->expect_new_payment_process_factor( Factor::WOOPAY_PAYMENT(), false );
+		$this->wcpay_gateway->should_use_new_process( $order );
+	}
+
+	/**
+	 * Testing the positive WCPay subscription signup factor is not possible,
+	 * as the check relies on the existence of the `WC_Subscriptions` class
+	 * through an un-mockable method, and the class simply exists.
+	 */
+	public function test_should_use_new_process_determines_negative_wcpay_subscription_signup() {
+		// The new payment process is only accessible in dev mode.
+		WC_Payments::mode()->dev();
+
+		$order = WC_Helper_Order::create_order();
+
+		WC_Subscriptions::$wcs_order_contains_subscription = '__return_true';
+		add_filter( 'wcpay_is_wcpay_subscriptions_enabled', '__return_true' );
+
+		$this->expect_new_payment_process_factor( Factor::WCPAY_SUBSCRIPTION_SIGNUP(), false );
+		$this->wcpay_gateway->should_use_new_process( $order );
+	}
+
+	public function test_new_process_payment() {
+		// The new payment process is only accessible in dev mode.
+		WC_Payments::mode()->dev();
+
+		$mock_service  = $this->createMock( PaymentProcessingService::class );
+		$mock_router   = $this->createMock( Router::class );
+		$order         = WC_Helper_Order::create_order();
+		$mock_response = [ 'success' => 'maybe' ];
+
+		wcpay_get_test_container()->replace( PaymentProcessingService::class, $mock_service );
+		wcpay_get_test_container()->replace( Router::class, $mock_router );
+
+		$mock_router->expects( $this->once() )
+			->method( 'should_use_new_payment_process' )
+			->willReturn( true );
+
+		// Assert: The new service is called.
+		$mock_service->expects( $this->once() )
+			->method( 'process_payment' )
+			->with( $order->get_id() )
+			->willReturn( $mock_response );
+
+		$result = $this->wcpay_gateway->process_payment( $order->get_id() );
+		$this->assertSame( $mock_response, $result );
+	}
+
+	/**
+	 * Sets up the expectation for a certain factor for the new payment
+	 * process to be either set or unset.
+	 *
+	 * @param string $factor_name Name of the factor.
+	 * @param bool   $value       Expected value.
+	 */
+	private function expect_new_payment_process_factor( $factor_name, $value ) {
```

Comment:
This took me a while to understand the purpose of tests calling out this private method. That's because it's a bit confusing between two similar method names in two classes: 

- WC_Payment_Gateway_WCPay::should_use_new_process
- Router::should_use_new_payment_process

Hence I suggest this name to describe better what it is going to do: 
 
```suggestion
	private function expect_router_should_use_new_payment_process_called_with_factor( $factor_name, $value ) {
```

Reply:
I understand the confusion, so I renamed the method. However, `expect_router_should_use_new_payment_process_called_with_factor` seemed needlessly long, especially considering the router has a single significant public method, which is why I switched to `expect_router_factor`. Hope it helps :)